util/extsort/disk_sorter.go

```diff
@@ -80,37 +746,123 @@ func OpenDiskSorter(dirname string, opts *DiskSorterOptions) (*DiskSorter, error
+	readerPool := newSSTReaderPool(fs, dirname, cache)
+	d := &DiskSorter{
+		opts:       opts,
+		fs:         fs,
+		dirname:    dirname,
+		cache:      cache,
+		readerPool: readerPool,
+	}
+	if err := d.init(); err != nil {
+		return nil, err
+	}
+	return d, nil
+}
 
-	dbDir := filepath.Join(dirname, "db")
-	tmpDir := filepath.Join(dirname, "tmp")
+func (d *DiskSorter) init() error {
+	list, err := d.fs.List(d.dirname)
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-	// Clean up the temporary directory.
-	if err := dbOpts.FS.RemoveAll(tmpDir); err != nil {
-		return nil, errors.Trace(err)
+	g, ctx := errgroup.WithContext(context.Background())
+	outputCh := make(chan *fileMetadata, len(list))
+	for _, name := range list {
+		if strings.HasSuffix(name, tmpFileSuffix) {
+			_ = d.fs.Remove(d.fs.PathJoin(d.dirname, name))
+			continue
+		}
+		fileNum, ok := parseFilename(d.fs, name)
+		if !ok {
+			continue
+		}
+		if int64(fileNum) > d.idAlloc.Load() {
+			d.idAlloc.Store(int64(fileNum))
+		}
+		g.Go(func() error {
+			if ctx.Err() != nil {
+				return errors.Trace(ctx.Err())
+			}
+			file, err := d.readFileMetadata(fileNum)
+			if err != nil {
+				return err
+			}
+			outputCh <- file
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		return err
 	}
-	if err := dbOpts.FS.MkdirAll(tmpDir, 0755); err != nil {
+
+	close(outputCh)
+	files := make([]*fileMetadata, 0, len(outputCh))
+	for file := range outputCh {
+		files = append(files, file)
+	}
+	if _, err := d.fs.Stat(d.fs.PathJoin(d.dirname, diskSorterSortedFile)); err == nil {
+		slices.SortFunc(files, func(a, b *fileMetadata) bool {
+			return bytes.Compare(a.startKey, b.startKey) < 0
+		})
+		d.orderedFiles = files
+		d.state.Store(diskSorterStateSorted)
+	} else {
+		d.pendingFiles.files = files
+		d.state.Store(diskSorterStateWriting)
+	}
+	return nil
+}
+
+func (d *DiskSorter) readFileMetadata(fileNum int) (*fileMetadata, error) {
+	reader, err := d.readerPool.get(fileNum)
+	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	defer func() {
+		_ = d.readerPool.unref(fileNum)
+	}()
 
-	db, err := pebble.Open(dbDir, dbOpts)
+	meta := &fileMetadata{
+		fileNum: fileNum,
+	}
+	if prop, ok := reader.Properties.UserProperties[kvStatsPropKey]; ok {
+		if err := json.Unmarshal([]byte(prop), &meta.kvStats); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	iter, err := reader.NewIter(nil, nil)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	return &DiskSorter{
-		db:     db,
-		dbOpts: dbOpts,
-		opts:   opts,
-		dbDir:  dbDir,
-		tmpDir: tmpDir,
-		idGen:  new(atomic.Int64),
-	}, nil
+	defer func() {
+		_ = iter.Close()
+	}()
+
+	firstKey, _ := iter.First()
+	if firstKey != nil {
+		meta.startKey = slices.Clone(firstKey.UserKey)
+	} else if err := iter.Error(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	lastKey, _ := iter.Last()
+	if lastKey != nil {
+		meta.lastKey = slices.Clone(lastKey.UserKey)
+		meta.endKey = append(meta.lastKey, 0)
```

Comment:
I'm not sure `slices.Clone` can guarantee to return a slice with no free capacity so `append` will not change the slice in place 🤔

Reply:
meta is read-only. It does not matter whether they share the same underlying slice.

https://go.dev/play/p/8hpvjZqHC5n